### ISSUE TITLE
False positive non-optimal if conditions

### ISSUE THREAD

**Issue by torinaki** [kalessil/phpinspectionsea#748]
In the following example, we are using short-circuit evaluation to call `isApplicableSpecificField` idempotent method. It will be called only once. But we get a warning that "if condition can be optimized".
![image](https://user-images.githubusercontent.com/1269372/34954318-9dc30ace-fa28-11e7-87a2-b4bb2db5b189.png)

To avoid this warning we can refactor code and use a temp variable. The difference is that in the case when 5th or 6th conditions is true then `isApplicableSpecificField` method call is excessive.
```php
        foreach ($where as $field => $value) {
            $isApplicableSpecificField = $this->isApplicableSpecificField($field);
            if ($field === 'range' && $isApplicableSpecificField) {  // condition 1
                // some job
            } elseif ($field === 'excludeIds' && $isApplicableSpecificField) { // condition 2
                // some job
            } elseif ($field === 'recommend' && $isApplicableSpecificField) { // condition 3
                // some job
            } elseif ($field === 'author.expert' && $isApplicableSpecificField) { // condition 4
                // some job
            } elseif (is_array($value)) { // condition 5
                // some job
            } else { // condition 6
                // some job
            }
        }
```

**Comment by rentalhost** [comment]
You too can refactorate the code like that:

```php
if (is_array($value)) {
}
elseif ($this->isApplicableSpecificField($field)) {
    switch ($field) {
    }
}
else {
}
```

By other side, I think that is valid your case while `&& (last condition)` will always run the last condition if previous is true, so it will be runned once in all case.

**Comment by kalessil** [comment]
I should admit you are on the right track when looking at the direction of refactoring (technically the inspection did its job). This might be complicated, but I think it worth brainstorming an alternative approach with your teammates.

**Comment by torinaki** [comment]
@rentalhost, @kalessil thanks you for the quick response. 

In case of `// condition 6` we will call `isApplicableSpecificField` excessively. Our goal is to avoid this call.

Refactoring option that you have provided have different logic. How exactly I need to handle `switch{ default: ... }` case? In my original implementation, it was `// condition 6`. I can copy/paste code from `// condition 6`, but its rise different issue and is not an option. And still, we have `isApplicableSpecificField` excessive call.

**Comment by kalessil** [comment]
```php
$specualFields = [...];
foreach ($where as $field => $value) {
           if (is_string($value) && in_array($value, $specualFields, true) && $this->isApplicableSpecificField($field)) {
                $this->invokeJobForField($value);
           } elseif (is_array($value)) {
                $this->invokeJobForArray($value);
           } else {
                $this->invokeDefaulrJob($value);
           }
        }
```

**Comment by torinaki** [comment]
@kalessil I assume that I still will need another if/else block inside `invokeJobForField` to execute exact logic base on `$field` name like following:
```php
...
    if (is_string($value) && in_array($value, $specialFields, true) && $this->isApplicableSpecificField($field)) {
            if ($field === 'range') {  // condition 1
                // some job
            } elseif ($field === 'excludeIds') { // condition 2
                // some job
            } elseif ($field === 'recommend') { // condition 3
                // some job
            } elseif ($field === 'author.expert') { // condition 4
                // some job
            } else {
                // throw exception
            }
    }
...
```
So I need to duplicate `in_array($value, $specialFields, true)` condition.

Thank you for your time. I more than sure that it is not the first priority. So let do not spend more time on this. If someone will have similar issue and thoughts how to properly handle it then he/she can reopen this ticket.

**Comment by kalessil** [comment]
@torinaki : got it. Agree, let see if we'll get similar reports.
